scrapy_poet/page_input_providers.py

```diff
@@ -54,49 +55,28 @@ def __init__(self, response: Response):
         """
 
     @abc.abstractmethod
-    def __call__(self):
+    def __call__(self, provided_classes: typing.Set[typing.Type]) -> typing.Dict[typing.Type, typing.Any]:
         """This method is responsible for building Page Input dependencies."""
 
+    @classmethod
+    def register(cls):
+        providers.add(cls)
 
-def register(provider_class: typing.Type[PageObjectInputProvider],
-             provided_class: typing.Type):
-    """This method registers a Page Object Input Provider in the providers
-    registry. It could be replaced by the use of the ``provides`` decorator.
-
-    Example:
-
-        register(ResponseDataProvider, ResponseData)
-    """
-    providers[provided_class] = provider_class
-
-
-def provides(provided_class: typing.Type):
-    """This decorator should be used with classes that inherits from
-    ``PageObjectInputProvider`` in order to automatically register them as
-    providers.
 
-    See ``ResponseDataProvider``'s implementation for an example.
-    """
-    def decorator(provider_class: typing.Type[PageObjectInputProvider]):
-        register(provider_class, provided_class)
-        return provider_class
-
-    return decorator
-
-
-@provides(ResponseData)
 class ResponseDataProvider(PageObjectInputProvider):
     """This class provides ``web_poet.page_inputs.ResponseData`` instances."""
 
+    provided_classes = {ResponseData}
+
     def __init__(self, response: Response):
         """This class receives a Scrapy ``Response`` as a dependency."""
         self.response = response
 
-    def __call__(self):
-        """This method builds a ``ResponseData`` instance using a Scrapy
-        ``Response``.
-        """
-        return ResponseData(
-            url=self.response.url,
-            html=self.response.text
-        )
+    def __call__(self, provided_classes: typing.Set[typing.Type]):
+        """Builds a ``ResponseData`` instance using a Scrapy ``Response``."""
+        return {
+            ResponseData: ResponseData(
+                url=self.response.url,
+                html=self.response.text
+            )
+        }
```

Comment:
You may think that we should:

- check if `provided_classes` are contained in `self.provided_classes` but our injection middleware already takes care of this so this method shouldn't be called with classes that it does not provide
- return a blank dictionary if there's no class being specified but again, our injection middleware already takes care of this so no provider should be initialized nor called if it doesn't provide any of the required types